let apiserver support api discovery

caesarxuchao opened the pull request:
The related proposal is here: https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/api-group.md#server-side-implementation

at apis/ (or api/), all supported groups and their versions are returned, in the form of "group/version", encoded in api.APIVersions;
![image](https://cloud.githubusercontent.com/assets/2823529/9867593/545efd34-5b24-11e5-8c1c-c116b2a726d0.png)

at apis/group (currently it's "experimental/", #13945 will make it "apis/experiemental"), all supported versions are returned, in the form of "group/version", encoded in api.APIVersions;
![image](https://cloud.githubusercontent.com/assets/2823529/9867595/563f028e-5b24-11e5-92e8-67d73235f7ae.png)

at apis/group/version, names of supported resources and whether the resource is namespaced is returned, encoded in api.APIResourceList;
![image](https://cloud.githubusercontent.com/assets/2823529/9867711/b522d5fe-5b25-11e5-86ac-8b31c0bc90bd.png)

@bgrant0607 as I added types in unversioned.go
@lavalamp @nikhiljindal

Commit:
let apiserver support api discovery

ncdc commented:
@kubernetes/rh-cluster-infra @kubernetes/rh-ux

bgrant0607 commented:
cc @smarterclayton

bgrant0607 commented on pkg/api/unversioned.go:
```diff
@@ -29,6 +29,19 @@ type APIVersions struct {
 
+// APIResource stores the name of a resource and whether it is namespaced.
+type APIResource struct {
+	Resource   string `json:"resource"`
```
Please document all fields.

smarterclayton replied:
I'm wondering whether we should expose "kind" for both request and response now, or let that stay swagger.  I think we want to have input Kind so that a client can search for the resource to post an arbitrary object from create -f -.  Kind would only be set if this is the canonical "resource", not secondary resources that perform specific actions.

We should definitely resist duplicating swagger as much as possible (in the future we can link directly to the swagger doc for this resource endpoint).

smarterclayton commented on pkg/apiserver/api_installer.go:
```diff
@@ -97,7 +99,7 @@ func (a *APIInstaller) newWebService() *restful.WebService {
 }
 
-func (a *APIInstaller) registerResourceHandlers(path string, storage rest.Storage, ws *restful.WebService, proxyHandler http.Handler) error {
+func (a *APIInstaller) registerResourceHandlers(path string, storage rest.Storage, ws *restful.WebService, proxyHandler http.Handler) (api.APIResource, error) {
```
Return a pointer to api.Resource and make all the returns be nil.

smarterclayton commented on pkg/master/master.go:
```diff
@@ -96,6 +96,12 @@ const (
+var (
+	// allAPIVersions records all AVIVersions that are supported at /apis, in the
+	// form of "group/version"
+	allAPIVersions []string
```
Why is this a package level variable?

caesarxuchao replied:
Ah, I forgot to leave a note here. It's package level because we will need to add/delete elements in this variable when installing/deleting thirdparty resource, https://github.com/kubernetes/kubernetes/blob/master/pkg/master/master.go#L777

smarterclayton replied:
Isn't the third party resource part of Master?  Maybe I'm confused but the
scope should be the master object.

On Tue, Sep 15, 2015 at 12:49 PM, Chao Xu <notifications@github.com> wrote:

> In pkg/master/master.go
> <https://github.com/kubernetes/kubernetes/pull/13955#discussion_r39534611>
> :
>
> > @@ -96,6 +96,12 @@ const (
> >  	DefaultEtcdPathPrefix = "/registry"
> >  )
> >
> > +var (
> > +	// allAPIVersions records all AVIVersions that are supported at /apis, in the
> > +	// form of "group/version"
> > +	allAPIVersions []string
>
> Ah, I forgot to leave a note here. It's package level because we will need
> to add/delete elements in this variable when installing/deleting thirdparty
> resource,
> https://github.com/kubernetes/kubernetes/blob/master/pkg/master/master.go#L777
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13955/files#r39534611>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

lavalamp replied:
Don't use a global for this; at run time you should just get the list from Brendan's 3rd party resource controller thingy.

lavalamp replied:
at a minimum this needs a lock.

lavalamp commented on pkg/api/unversioned.go:
```diff
@@ -24,14 +24,33 @@ import (
+type APIResourceList struct {
+	// APIResources store the name of the resources and whether they are
+	// namespaced.
+	APIResources []APIResource `json:"resources"`
```
@bgrant0607 Would it be useful to add `Version string` as a field here? I know the client should already know that since it's in the URL path, but somehow it seems like it'd be nice to repeat it for them here.

Alternatively, it could go in the APIResource struct but that'd be very repetitive.

bgrant0607 replied:
Sounds reasonable. 

With or without the group?

caesarxuchao replied:
I think with the group is better. @bgrant0607, I'll add the field if you agree.

lavalamp commented on pkg/apiserver/api_installer.go:
```diff
@@ -78,11 +78,13 @@ func (a *APIInstaller) Install() (ws *restful.WebService, errors []error) {
+		if err != nil {
 			errors = append(errors, err)
 		}
+		apiResources = append(apiResources, *apiResource)
```
I don't think apiResources needs to be a global, especially not if you're returning it from this function.

caesarxuchao replied:
no, apiResources is not a global, the global one is allAPIVersions :)

lavalamp replied:
Oh right, sorry!

caesarxuchao commented on pkg/master/master.go:
```diff
@@ -570,16 +570,30 @@ func (m *Master) init(c *Config) {
 
+	// allAPIVersions records all AVIVersions that are supported at /apis, in the
+	// form of "group/version"
+	allAPIVersions := []string{}
```
@smarterclayton @lavalamp, I move allAPIVersions into init(). When Brendan's PR is in, we can pass allAPIVersions to 3rd party controller so it can add/delete versions from it.

smarterclayton replied:
When we init the third party controller, we should be able to fetch the
versions on the initial run and then remove the package variable in favor
of a member on Master?  If so I'm fine with that as long as you add a TODO
that this should be removed and we have a follow up.

On Tue, Sep 15, 2015 at 2:23 PM, Chao Xu <notifications@github.com> wrote:

> In pkg/master/master.go
> <https://github.com/kubernetes/kubernetes/pull/13955#discussion_r39546584>
> :
>
> > @@ -570,16 +570,30 @@ func (m *Master) init(c *Config) {
> >  	requestInfoResolver := &apiserver.APIRequestInfoResolver{APIPrefixes: sets.NewString(strings.TrimPrefix(defaultVersion.Root, "/")), RestMapper: defaultVersion.Mapper}
> >  	apiserver.InstallServiceErrorHandler(m.handlerContainer, requestInfoResolver, apiVersions)
> >
> > +	// allAPIVersions records all AVIVersions that are supported at /apis, in the
> > +	// form of "group/version"
> > +	allAPIVersions := []string{}
>
> @smarterclayton <https://github.com/smarterclayton> @lavalamp
> <https://github.com/lavalamp>, I move allAPIVersions into init(). When
> Brendan's PR is in, we can pass allAPIVersions to 3rd party controller so
> it can add/delete versions from it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13955/files#r39546584>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

smarterclayton replied:
Nm, I saw you fixed it.

caesarxuchao replied:
Hm, probably I should make it a field of Master, rather than a local variable.

bgrant0607 commented:
Proposal: https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/api-group.md#server-side-implementation

bgrant0607 commented on pkg/api/unversioned.go:
```diff
@@ -24,14 +24,33 @@ import (
 	Versions []string `json:"versions"`
 }
 
+// APIResource stores the name of a resource and whether it is namespaced.
```
s/stores/specifies/

bgrant0607 commented on pkg/api/unversioned.go:
```diff
@@ -24,14 +24,33 @@ import (
+	Namespaced bool `json:"namespaced"`
+}
+
+// APIResourceList is a list of APIResource, it is used to store the name of the
```
Don't use the word "store", here nor anywhere in the doc.

bgrant0607 commented on pkg/api/unversioned.go:
```diff
@@ -24,14 +24,33 @@ import (
+// resources supported in a specific group and version, and whether the resource
+// is namespaced.
+type APIResourceList struct {
+	// APIResources store the name of the resources and whether they are
```
Find an alternative to "store".

bgrant0607 commented on pkg/api/unversioned.go:
```diff
@@ -24,14 +24,33 @@ import (
+type APIResource struct {
+	// Resource is the name of the resource.
+	Resource string `json:"resource"`
+	// Namespaced indicates if a resource is namespaced or not.
```
s/if/whether/

bgrant0607 commented on pkg/api/unversioned.go:
```diff
@@ -24,14 +24,33 @@ import (
-// strings in order to allow for future evolution, though unversioned
+// strings in order to allow for future evolution, though unversioned.
 type APIVersions struct {
+	// Versions are the api versions that are available.
```
Aren't these the groups and versions?

caesarxuchao replied:
Yes, groups and versions. I will modify the comment.

caesarxuchao replied:
@bgrant0607, the commit "modify comments" fix the comment issues. PTAL. Thanks.

lavalamp commented:
@bgrant0607 @caesarxuchao @JanetKuo 

I want to raise the issue of whether the server should communicate a preferred version or not.

The use case is a client (e.g. kubectl) using a resource known to apiserver but not the client, so the client doesn't have preconceived notions about what version of the resource it would like to use (and the client is of course limited to completely generic operations on the resource).

Possible rules the client could use are:
* always use the most recent version
* always use the oldest version

But the server knows an additional piece of information, which is what version is being stored. If the server communicated this preferred version, clients could use it and we'd potentially be able to avoid unnecessary conversions.

Should we add this to one of these responses?

lavalamp commented on pkg/api/unversioned.go:
```diff
@@ -22,16 +22,35 @@ import (
-// strings in order to allow for future evolution, though unversioned
+// APIVersions lists the groups and versions that are available, to allow
+// clients to discover the API. APIVersions isn't just an unnamed array of
+// strings in order to allow for future evolution, though unversioned.
```
Maybe remove "APIVersions isn't just an unnamed array of strings in order to allow for future evolution, though unversioned."

It is very confusing.

caesarxuchao replied:
Removed. Thanks.

lavalamp commented on pkg/master/master.go:
```diff
@@ -570,16 +570,30 @@ func (m *Master) init(c *Config) {
 			glog.Fatalf("Unable to setup experimental api: %v", err)
 		}
-		apiserver.AddApiWebService(m.handlerContainer, c.ExpAPIPrefix, []string{expVersion.Version})
+		g, err := latest.Group("experimental")
```
I'd like to see a loop here that goes through every Group, instead of specifically referencing experimental. But we can fix this when we add additional groups, I guess.

caesarxuchao replied:
Could you explain why we need a loop here? This block is enclosed in "if m.exp", so we know it's for experimental.

smarterclayton replied:
We'll want to do that from Openshift, so we can add that when we get to
it.  Or you can add a field to MasterConfig now that is a list of external
groups.

On Sep 15, 2015, at 6:04 PM, Daniel Smith <notifications@github.com> wrote:

In pkg/master/master.go
<https://github.com/kubernetes/kubernetes/pull/13955#discussion_r39571743>:

>  	if m.exp {
>  		expVersion := m.experimental(c)
>  		if err := expVersion.InstallREST(m.handlerContainer); err != nil {
>  			glog.Fatalf("Unable to setup experimental api: %v", err)
>  		}
> -		apiserver.AddApiWebService(m.handlerContainer, c.ExpAPIPrefix, []string{expVersion.Version})
> +		g, err := latest.Group("experimental")

I'd like to see a loop here that goes through every Group, instead of
specifically referencing experimental. But we can fix this when we add
additional groups, I guess.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/13955/files#r39571743>.

lavalamp replied:
Yes, what Clayton says. We don't want to end up with a bunch of individually named booleans where we check each individually and then use the group of the same name, we want to loop over a single list.

caesarxuchao replied:
I see. I'll do it when we have multiple groups.

smarterclayton commented:
This also sounds like a potential attack vector, although that's unlikely.
It's a good performance optimization for when client and server can be
flexible, so I'd prefer to have it (as preferredVersion).  Generic clients
may want the newest version in practice because they can ensure users get
to choose the widest range of features to use (with edit and a new command,
more features will be visible to the user).

On Sep 15, 2015, at 5:55 PM, Daniel Smith <notifications@github.com> wrote:

@bgrant0607 <https://github.com/bgrant0607> @caesarxuchao
<https://github.com/caesarxuchao> @JanetKuo <https://github.com/JanetKuo>

I want to raise the issue of whether the server should communicate a
preferred version or not.

The use case is a client (e.g. kubectl) using a resource known to apiserver
but not the client, so the client doesn't have preconceived notions about
what version of the resource it would like to use (and the client is of
course limited to completely generic operations on the resource).

Possible rules the client could use are:

   - always use the most recent version
   - always use the oldest version

But the server knows an additional piece of information, which is what
version is being stored. If the server communicated this preferred version,
clients could use it and we'd potentially be able to avoid unnecessary
conversions.

Should we add this to one of these responses?

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/13955#issuecomment-140557885>
.

caesarxuchao commented:
I added groupVersion to APIResourcesList, also added preferredVersion to APIVersions.
After the change, 
at /apis/:
![image](https://cloud.githubusercontent.com/assets/2823529/9914067/f3490502-5c64-11e5-8f8a-2a0f32ee6c32.png)

at /apis/experimental:
![image](https://cloud.githubusercontent.com/assets/2823529/9914067/f3490502-5c64-11e5-8f8a-2a0f32ee6c32.png)

at /apis/experimental/v1/: (**the groupVersion only contains version, it will be fixed in another PR**)
![image](https://cloud.githubusercontent.com/assets/2823529/9914105/18e42bd4-5c65-11e5-8bc3-bb07926eb379.png)

smarterclayton commented:
I would expect `/apis/` to return groupVersion, group, and version.  If someone knows the group but not the version, how do they find it in the list?  They shouldn't have to do string manipulation.

caesarxuchao commented:
@smarterclayton, if they know the group, they can visit /apis/<group> to find the versions.

lavalamp commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -245,12 +245,13 @@ func (s *APIServer) verifyClusterIPFlags() {
 }
 
-func newEtcd(etcdConfigFile string, etcdServerList []string, interfacesFunc meta.VersionInterfacesFunc, defaultVersion, storageVersion, pathPrefix string) (etcdStorage storage.Interface, err error) {
+func newEtcd(etcdConfigFile string, etcdServerList []string, interfacesFunc meta.VersionInterfacesFunc, defaultVersion, storageVersion, pathPrefix string) (storage.Interface, string, error) {
```
Let's:
* Not remove the return parameter names; they help readability significantly.
* Not default storage version to default version; instead, return an error if storage version is ""
* Default version isn't even used and is confusing; remove it as a parameter

lavalamp commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -339,7 +341,13 @@ func (s *APIServer) Run(_ []string) error {
+	if err != nil {
+		return err
+	}
+	storageVersions := make(map[string]string)
```
This is good-- we should have a map of group to storage version. (can we call it groupToStorageVersion?)

In the near future, maybe not this PR:

We should deprecate StorageVersion and ExpStorageVersion, and replace them with StorageVersions, which should be a map from group to version (it can be specified in the form: "experimental/v1alpha1,core/v1,/v1")

newEtcd should then be called in a loop, etc. Do we really need to call newEtcd for every group?

caesarxuchao replied:
> Do we really need to call newEtcd for every group?

I think so. newEtcd take InterfacesFor as an argument, which is different for every group.

smarterclayton commented:
I want to visit /apis/ to find groups.  I won't know the groups by default.

On Wed, Sep 16, 2015 at 4:54 PM, Chao Xu <notifications@github.com> wrote:

> @smarterclayton <https://github.com/smarterclayton>, if they know the
> group, they can visit /apis/ to find the versions.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13955#issuecomment-140891567>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

caesarxuchao commented:
@smarterclayton, is this what you are suggesting: instead of visiting /apis to find group, and then visiting /apis/<group> to find versions for a group, clients should be able to discover all the information at /apis/, including all supported"group/version", all groups, and lists of supported versions in each group?

smarterclayton commented:
Yes

On Wed, Sep 16, 2015 at 5:13 PM, Chao Xu <notifications@github.com> wrote:

> @smarterclayton <https://github.com/smarterclayton>, is this what you are
> suggesting: instead of visiting /apis to find group, and then visiting
> /apis/ to find versions for a group, clients should be able to discover all
> the information at /apis/, including all supported"group/version", all
> groups, and lists of supported versions in each group?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13955#issuecomment-140896785>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

liggitt commented:
As someone who will have to write a discovering client, I wholeheartedly agree with not wanting to do N requests to discover groups, versions, and resources

caesarxuchao commented:
@smarterclayton, @lavalamp, @liggitt, I just pushed a new commit to implement Clayton's comment. PTAL.

At /apis, assuming we also supports another group "group2",
```
{
  "versions": [
    "experimental/v1"
    "group2/v2"
    "group2/v3"
  ],
  "groups": [
    "experimental"
    "group2"
  ],
  "groupMap": {
    "experimental": {
      "group": "experimental",
      "versions": [
        "v1"
      ],
      "PreferredVersion": "v1"
    },
    "group2": {
      "group": "group2",
      "versions": [
        "v2",
        "v3",
      ],
      "PreferredVersion": "v2"
    }
  }
}
```

caesarxuchao commented:
This is what returned at apis/experimental
```
{
  "group": "experimental",
  "versions": [
    "v1"
  ],
  "PreferredVersion": "v1"
}
```

smarterclayton commented:
It would probably be more canonical to have

// Top level
groups []group

group struct {
  name string
  versions []version
}

version struct {
  name string
}

Because these are outside of an API version we need to be able to grow them
without changing - I expect we will add url or uri to both version and
group in the future.  And possibly other fields.

The map doesn't really fit our existing patterns


On Sep 16, 2015, at 6:57 PM, Chao Xu <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton>, @lavalamp
<https://github.com/lavalamp>, @liggitt <https://github.com/liggitt>, I
just pushed a new commit to implement Clayton's comment. PTAL.

At /apis, assuming we also supports another group "group2",

{
  "versions": [
    "experimental/v1"
    "group2/v2"
    "group2/v3"
  ],
  "groups": [
    "experimental"
    "group2"
  ],
  "groupMap": {
    "experimental": {
      "group": "experimental",
      "versions": [
        "v1"
      ],
      "PreferredVersion": "v1"
    },
    "group2": {
      "group": "group2",
      "versions": [
        "v2",
        "v3",
      ],
      "PreferredVersion": "v2"
    }
  }
}

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/13955#issuecomment-140919366>
.

caesarxuchao commented:
@smarterclayton, I agree.

I was trying to re-use the APIVersions struct which is return at /api. I think I'll just create bunch of new structs as you suggested for /apis, and only use the old APIVersions at /api.

lavalamp commented:
If we're making a new struct, then please call it groupVersions if you're printing things of the form group/version. This is going to confuse everyone :(

smarterclayton commented:
In my head it was groups, then nested inside versions.  No top level
"groupversions".  The nested versions inside could be "groupversions"

On Wed, Sep 16, 2015 at 7:24 PM, Daniel Smith <notifications@github.com>
wrote:

> If we're making a new struct, then please call it groupVersions if you're
> printing things of the form group/version. This is going to confuse
> everyone :(
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13955#issuecomment-140924613>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

lavalamp commented:
Yeah, I prefer it denormalized, too. Like:

```go
type Group struct {
  // e.g. "experimental"
  GroupName string
  // e.g. "v1", "v2"...
  Versions []string
  PreferredVersion string
}

type Groups struct {
  Groups []Group
}
```

We should default to making group and version separate fields whenever possible. When we stick them together, we should call it groupVersion whenever possible. We break these rules ONLY when backwards compatibility requires us to.

smarterclayton commented:
We should probably include groupVersion at the deepest level so clients
aren't doing string concatenation and parsing.

On Sep 16, 2015, at 7:48 PM, Daniel Smith <notifications@github.com> wrote:

Yeah, I prefer it denormalized, too. Like:

type Group struct {
  // e.g. "experimental"
  GroupName string
  // e.g. "v1", "v2"...
  Versions []string
  PreferredVersion string
}
type Groups struct {
  Groups []Group
}

We should default to making group and version separate fields whenever
possible. When we stick them together, we should call it groupVersion
whenever possible. We break these rules ONLY when backwards compatibility
requires us to.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/13955#issuecomment-140927465>
.

caesarxuchao commented:
Here's the output in the newest commit:
at apis/
```
{
  "groups": [
    {
      "groupName": "experimental",
      "versions": [
        {
          "version": "v1"
        }
      ],
      "PreferredVersion": {
        "version": "v1"
      }
    }
  ]
}
```
at /apis/experimental:
```
{
  "groupName": "experimental",
  "versions": [
    {
      "version": "v1"
    }
  ],
  "PreferredVersion": {
    "version": "v1"
  }
}
```
and the legacy behavior at /api:
```
{
  "versions": [
    "v1"
  ]
}
```

caesarxuchao commented:
@lavalamp , sorry I haven't added the comments for the new types yet, will do that in another commit.

lavalamp commented:
While we're making you jump through hoops, can you add groupVersions in addition to versions?

lavalamp commented on pkg/api/unversioned.go:
```diff
@@ -22,16 +22,49 @@ import (
+
+type APIGroup struct {
+	GroupName        string    `json:"groupName"`
+	Versions         []Version `json:"versions"`
```
I think this can just be an array of strings?

caesarxuchao replied:
Quoting @smarterclayton: "I expect we will add url or uri to both version and
group in the future"

So we should use Version struct to keep the extensibility.

caesarxuchao replied:
Also, following this logic, should we also make GroupVersion a struct rather than an array of string?

smarterclayton replied:
For Openshift I'd like to embed uris here as soon as we sort out component
registration so we can do relative or absolute pointers from versions to
target (and same for group)

On Sep 16, 2015, at 8:21 PM, Chao Xu <notifications@github.com> wrote:

In pkg/api/unversioned.go
<https://github.com/kubernetes/kubernetes/pull/13955#discussion_r39700847>:

>  	Versions []string `json:"versions"`
>  }
>
> +type APIGroupList struct {
> +	Groups []APIGroup `json:"groups"`
> +}
> +
> +type APIGroup struct {
> +	GroupName        string    `json:"groupName"`
> +	Versions         []Version `json:"versions"`

Quoting @smarterclayton <https://github.com/smarterclayton>: "I expect we
will add url or uri to both version and
group in the future"

So we should use Version struct to keep the extensibility.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/13955/files#r39700847>.

lavalamp replied:
Ah-- in that case, I'd add GroupVersion to the Version struct directly, since the URL is the same.

smarterclayton replied:
Alternatively pod groupVersion string inside version struct, and client can
scan it to find their group version (secondary index)

On Sep 16, 2015, at 8:23 PM, Chao Xu <notifications@github.com> wrote:

In pkg/api/unversioned.go
<https://github.com/kubernetes/kubernetes/pull/13955#discussion_r39700982>:

>  	Versions []string `json:"versions"`
>  }
>
> +type APIGroupList struct {
> +	Groups []APIGroup `json:"groups"`
> +}
> +
> +type APIGroup struct {
> +	GroupName        string    `json:"groupName"`
> +	Versions         []Version `json:"versions"`

Also, following this logic, should we also make GroupVersion a struct
rather than an array of string?

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/13955/files#r39700982>.

smarterclayton commented:
Design by code review committee is more fun

On Sep 16, 2015, at 8:17 PM, Daniel Smith <notifications@github.com> wrote:

While we're making you jump through hoops, can you add groupVersions in
addition to versions?

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/13955#issuecomment-140930966>
.

lavalamp commented:
>Design by code review committee is more fun

If we're going to spend time getting something right, unversioned APIs are a good candidate!

smarterclayton commented:
Real developers version their Apis in production.

On Sep 16, 2015, at 8:26 PM, Daniel Smith <notifications@github.com> wrote:

Design by code review committee is more fun

If we're going to spend time getting something right, unversioned APIs are
a good candidate!

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/13955#issuecomment-140931949>
.

bgrant0607 commented:
Re. preferred version: Someone should look at the version negotiation code in kubectl.

caesarxuchao commented:
new version pushed.

at /apis:
```
{
  "groups": [
    {
      "groupName": "experimental",
      "groupVersions": [
        {
          "groupVersion": "experimental/v1",
          "version": "v1"
        }
      ],
      "PreferredVersion": {
        "groupVersion": "experimental/v1",
        "version": "v1"
      }
    }
  ]
}
```
at /apis/experimental:
```
{
  "groupName": "experimental",
  "groupVersions": [
    {
      "groupVersion": "experimental/v1",
      "version": "v1"
    }
  ],
  "PreferredVersion": {
    "groupVersion": "experimental/v1",
    "version": "v1"
  }
}
```

caesarxuchao commented:
Hi everyone, is there more comment? 

Now at /apis:
```
{
  "groups": [
    {
      "groupName": "experimental",
      "groupVersions": [
        {
          "groupVersion": "experimental/v1",
          "version": "v1"
        }
      ],
      "PreferredVersion": {
        "groupVersion": "experimental/v1",
        "version": "v1"
      }
    }
  ]
}
```
at /apis/experimental:
```
{
  "groupName": "experimental",
  "groupVersions": [
    {
      "groupVersion": "experimental/v1",
      "version": "v1"
    }
  ],
  "PreferredVersion": {
    "groupVersion": "experimental/v1",
    "version": "v1"
  }
}
```

at /apis/experimental/v1,
```
{
  "groupVersion": "v1",
  "resources": [
    {
      "resource": "daemonsets",
      "namespaced": true
    },
    {
      "resource": "deployments",
      "namespaced": true
    },
    {
      "resource": "horizontalpodautoscalers",
      "namespaced": true
    },
    {
      "resource": "jobs",
      "namespaced": true
    },
    {
      "resource": "replicationcontrollers",
      "namespaced": true
    },
    {
      "resource": "replicationcontrollers/scale",
      "namespaced": true
    },
    {
      "resource": "thirdpartyresources",
      "namespaced": true
    }
  ]
}
```

bgrant0607 commented on pkg/api/unversioned.go:
```diff
@@ -22,16 +22,64 @@ import (
+	GroupVersions []GroupVersion `json:"groupVersions"`
+	// PreferredVersion is the version preferred by the API server, which
+	// probably is the storage version.
+	PreferredVersion GroupVersion `json: "preferredVersion"`
```
Can't be a space after `json:`

bgrant0607 commented on pkg/api/unversioned.go:
```diff
@@ -22,16 +22,64 @@ import (
-// version negotiation. APIVersions isn't just an unnamed array of
-// strings in order to allow for future evolution, though unversioned
+// APIVersions lists the versions that are available, to allow clients to
+// discover the API at /api.
```
Some comment about /api being the parent of the legacy v1 API would be useful.

bgrant0607 commented:
Note: capitalization of "PreferredVersion" above is a bug. It should be "preferredVersion". It appears to be correct in the code.

bgrant0607 commented:
Poll of people interested in naming:

What connotation does "API" (just "API") have to you, or at least what is the closest match among the concepts we now have? The entire Kubernetes API surface? A group? A group/version? A group/version/kind?

bgrant0607 commented on pkg/api/unversioned.go:
```diff
@@ -22,16 +22,64 @@ import (
+// of a group.
+type APIGroup struct {
+	// GroupName represents the name of the group.
+	GroupName string `json:"groupName"`
```
Our convention is to take advantage of context. Since this is an APIGroup, "group" is redundant in the field name. If we were to stick to this convention, this field would be called simply `name`. Did someone feel that was ambiguous in this situation? If so, was it in part because this type doesn't have a `kind` field?

lavalamp replied:
No, I just missed this. It should definitely just be "name".

liggitt replied:
Same for groupVersions > versions?

lavalamp replied:
No. Fields that have stuff in the form group/version should be named groupVersion, unless we're maintaining compatibility.

liggitt replied:
I was referring to the `groupVersions` field which is an array of version structs. Within a particular version struct is a `groupVersion` field in the form `<group>/<version>`, which I agree we should keep named that way

bgrant0607 commented on pkg/api/unversioned.go:
```diff
@@ -22,16 +22,64 @@ import (
+type GroupVersion struct {
+	// GroupVersion is in the form of "group/version".
+	GroupVersion string `json:"groupVersion"`
+	// Version is in the form of "version", this is to save the clients the
```
Break this into 2 sentences: This is to save...

bgrant0607 commented on pkg/api/unversioned.go:
```diff
@@ -22,16 +22,64 @@ import (
+// GroupVersion contains the "group/version" and "version" string of a version.
+// It is made a struct to keep extensiblity.
+type GroupVersion struct {
+	// GroupVersion is in the form of "group/version".
```
These comments (both fields) don't say what the fields are.

GroupVersion specifies the API group and version in the form "group/version".

bgrant0607 replied:
We should also mention that this is the form specified in the `apiVersion` field of resources in the APIs.

This made me consider whether this field should also be called `apiVersion`, but I think that would be more confusing. I also considered `versionWithGroup`, but the verbosity doesn't seem worth it and the existing form has the additional advantage of mimicking the form of the values, so `groupVersion` is fine.

bgrant0607 commented on pkg/api/unversioned.go:
```diff
@@ -22,16 +22,64 @@ import (
+	Version string `json:"version"`
+}
+
+// APIResource specifies the name of a resource and if it is namespaced.
```
s/if/whether/

bgrant0607 replied:
APPLIES TO ALL COMMENTS IN THIS FILE:

We're going to want to generate user-facing documentation from these comments. The Go types/fields are not a user-visible part of the API. Use the json/URL forms of the names, or just use plain English instead.

caesarxuchao replied:
I see. We need to fix comments in various types.go, too. Now they all use field/type names.

caesarxuchao replied:
Type names are actually visible in our doc. Field names are not.

lavalamp replied:
shouldn't the json:"foo" field names show?

caesarxuchao replied:
Let me clarify:
```
type Bar struct {
    Foo string `json:"foo"`
}
```
will show up in our doc as:
**Bar**:

Name | Description
---------|------------------
foo | comments

So the comment should be:
```
// Bar is ...
type Bar struct {
    // foo is ...
    Foo string `json:"foo"`
}
```

bgrant0607 replied:
We can leave the type names as is for now, but it's a bug, not a feature, that they are exposed in our docs and tools. We do not want them to be part of our API surface, with the same backward-compatibility requirements, etc.

smarterclayton commented:
I'm starting to think this should be a Boolean in the version struct.  I
don't think a client selects a preferred version without identifying the
group first.

On Sep 17, 2015, at 8:07 PM, Brian Grant <notifications@github.com> wrote:

Note: capitalization of "PreferredVersion" above is a bug. It should be
"preferredVersion". It appears to be correct in the code.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/13955#issuecomment-141291961>
.

smarterclayton commented:
A group probably.

On Sep 17, 2015, at 8:12 PM, Brian Grant <notifications@github.com> wrote:

Poll of people interested in naming:

What connotation does "API" (just "API") have to you, or at least what is
the closest match among the concepts we now have? The entire Kubernetes API
surface? A group? A group/version? A group/version/kind?

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/13955#issuecomment-141292536>
.

bgrant0607 commented:
@smarterclayton I agree. 

I briefly entertained replacing "group" with "api" in these types, but the ambiguity outweighs the pithiness. In docs and code, we should be careful to not use "API" in an ambiguous fashion.

bgrant0607 commented:
Re. preferred as a bool: That would match `namespaced`, as well. The advantages I see in the current form are:
1. Not possible to specify more than one version per group
2. No need to search the list

bgrant0607 commented on pkg/api/unversioned.go:
```diff
@@ -22,16 +22,64 @@ import (
+// APIResource specifies the name of a resource and if it is namespaced.
+type APIResource struct {
+	// Resource is the name of the resource.
+	Resource string `json:"resource"`
```
"Resource" seems ambiguous. Is that a path/link? Or a specific resource?

I'd use "ResourceName" to match the "GroupName" pattern, if we're sticking to that, or just "name" if we add a kind field.

bgrant0607 replied:
See the last bullet I recently added to the naming conventions:
https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md#naming-conventions

lavalamp replied:
Should be "Name".

bgrant0607 commented on pkg/api/unversioned.go:
```diff
@@ -22,16 +22,64 @@ import (
+type APIResource struct {
+	// Resource is the name of the resource.
+	Resource string `json:"resource"`
+	// Namespaced indicates if a resource is namespaced or not.
```
s/if/whether/

bgrant0607 commented:
I see from email that @smarterclayton previously brought up the issue of a `kind` field. It looks like that discussion didn't go anywhere? Was that `kind` for these new API types, or `kind` for resources discovered via these APIs?

bgrant0607 commented on pkg/api/unversioned.go:
```diff
@@ -22,16 +22,64 @@ import (
+}
+
 // RootPaths lists the paths available at root.
 // For example: "/healthz", "/api".
```
Replace /api with /apis

bgrant0607 replied:
Replace /api with /apis

lavalamp commented:
I also thought about making preferred a bool. I don't think either is inherently better, and would tend to go with the one that produced less duplication (and smaller JSON payloads). Now that I think about it again, I note we could omit all of the `"preferred": false` lines, so I guess this means I prefer it as a bool.

But is worth making clients search the list? I'm not sure.

liggitt commented:
> preferred as a bool: That would match namespaced, as well. The advantages I see in the current form are:
> 1. Not possible to specify more than one version per group
> 2. No need to search the list

I'm not that concerned with 1. For 2, any client hoping to use this discovery doc without searching lists is already going to be disappointed.

smarterclayton commented:
@jwforres as a consumer of this can you comment on the things you'd expect in a good api discovery doc?  Anything we've missed?  Anything we need to leave open for future extension?

liggitt commented:
Auth methods/endpoints come to mind, though that opens another can of worms (different auth per API group is possible, API server might not know auth endpoints in case of auth proxy, etc). That might be better handled as part of a 401 response.

smarterclayton commented:
Yeah I like 401 on fetching the API doc or an individual resource.

On Fri, Sep 18, 2015 at 11:28 AM, Jordan Liggitt <notifications@github.com>
wrote:

> Auth methods/endpoints come to mind, though that opens another can of
> worms (different auth per API group is possible, API server might not know
> auth endpoints in case of auth proxy, etc). That might be better handled as
> part of a 401 response.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13955#issuecomment-141482978>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

caesarxuchao commented:
Pushed a commit to address most of the comments. I still keep preferredVersion a separate field to save the client from searching a list. I can make a change if reviewers reach consensus.

liggitt commented on pkg/api/unversioned.go:
```diff
@@ -22,16 +22,64 @@ import (
+	// groupVersions are the versions supported in this group.
+	GroupVersions []GroupVersion `json:"groupVersions"`
+	// preferredVersion is the version preferred by the API server, which
+	// probably is the storage version.
```
is this guaranteed to be populated, or is preferredVersion optional?

caesarxuchao replied:
Thanks. I think we should make it optional. @lavalamp do you agree?

caesarxuchao commented on pkg/api/unversioned.go:
```diff
@@ -22,16 +22,64 @@ import (
+	// name is the name of the group.
+	Name string `json:"name"`
+	// groupVersions are the versions supported in this group.
+	GroupVersions []GroupVersion `json:"groupVersions"`
```
@liggitt, I think you are arguing this field should be named Versions? But it's a list of GroupVersion struct, so I think GroupVersions is a better name.

liggitt replied:
context rules... `"versions":[...]` within a group struct are understood to be the versions for the group, I think. If we were consistent, would you also say "preferredGroupVersion"? Seems redundant. I like keeping the `groupVersion` terminology for the string fields where we are combining the group and version name

caesarxuchao replied:
Yes, you convinced me. The "preferredGroupVersion" sounds redundant. For consistency, I will rename this field to Versions.

caesarxuchao replied:
Done.

liggitt commented:
Are groupVersion resources included in the main discovery doc now, or are additional requests still required to get the list of resources for each groupVersion

smarterclayton commented:
We may want to hold off on that until we know how we would surface the
Openshift group resources in this API remotely (cache that looks up
annotations on a service?)

On Sep 18, 2015, at 4:57 PM, Jordan Liggitt <notifications@github.com>
wrote:

Are groupVersion resources included in the main discovery doc now, or are
additional requests still required to get the list of resources for each
groupVersion

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/13955#issuecomment-141567346>
.

caesarxuchao commented:
No, we don't show the list of resources at the main doc (e.g. /apis)

caesarxuchao commented:
@smarterclayton, shall I go ahead to rebase and squash? Thanks.

smarterclayton commented:
I think this is LGTM, Brian needs to give final sign off on API.  Rebase and squash please.

caesarxuchao commented:
@bgrant0607 , squashed and rebased. Could you take another pass over the API? Thanks.

bgrant0607 commented:
LGTM

caesarxuchao commented:
Broke a test after rebase. Now fixed.

caesarxuchao commented:
Integration test is broken after rebase. Now fixed.

caesarxuchao commented:
reapply the LGTM as it's just rebase and nit fix.
